File: redome-front/src/app/shared/dto/prescricao.dto.ts
import { BaseEntidade } from "../base.entidade";
import { ConvertUtil } from "../util/convert.util";
import { ArquivoPrescricaoDTO } from "./arquivo.prescricao.dto";
import { PrescricaoCordaoDTO } from "./prescricao.cordao.dto";
import { PrescricaoMedulaDTO } from "./prescricao.medula.dto";

export class PrescricaoDTO extends BaseEntidade {

   private _id: number;
   private _idTipoPrescricao: number;
   private _idDoador: number;
   private _idTipoDoador: number;
   private _rmr: number;
   private _idCentroTransplante: number;
   private _medula: PrescricaoMedulaDTO;
   private _cordao: PrescricaoCordaoDTO;
   private _arquivosPrescricao: ArquivoPrescricaoDTO[] = [];
   private _arquivosPrescricaoJustificativa: ArquivoPrescricaoDTO[] = [];

   get id(): number {
      return this._id;
   }

   set id(value: number) {
      this._id = value;
   }

   get idTipoPrescricao(): number {
      return this._idTipoPrescricao;
   }

   set idTipoPrescricao(value: number) {
      this._idTipoPrescricao = value;
   }

   get idDoador(): number {
      return this._idDoador;
   }

   set idDoador(value: number) {
      this._idDoador = value;
   }


   get idTipoDoador(): number {
      return this._idTipoDoador;
   }

   set idTipoDoador(value: number) {
      this._idTipoDoador = value;
   }

   get rmr(): number {
      return this._rmr;
   }

   set rmr(value: number) {
      this._rmr = value;
   }

   get idCentroTransplante(): number {
      return this._idCentroTransplante;
   }

   set idCentroTransplante(value: number) {
      this._idCentroTransplante = value;
   }

   get medula(): PrescricaoMedulaDTO {
      return this._medula;
   }

   set medula(value: PrescricaoMedulaDTO) {
      this._medula = value;
   }

   get cordao(): PrescricaoCordaoDTO {
      return this._cordao;
   }

   set cordao(value: PrescricaoCordaoDTO) {
      this._cordao = value;
   }

   public get arquivosPrescricao(): ArquivoPrescricaoDTO[] {
		return this._arquivosPrescricao;
	}

	public set arquivosPrescricao(value: ArquivoPrescricaoDTO[]) {
		this._arquivosPrescricao = value;
	}

	public get arquivosPrescricaoJustificativa(): ArquivoPrescricaoDTO[] {
		return this._arquivosPrescricaoJustificativa;
	}

	public set arquivosPrescricaoJustificativa(value: ArquivoPrescricaoDTO[]) {
		this._arquivosPrescricaoJustificativa = value;
	}

	public jsonToEntity(res: any): PrescricaoDTO {

		if (res.arquivosPrescricao) {
			this.arquivosPrescricao = [];
			res.arquivosPrescricao.forEach(arquivo => {
				this.arquivosPrescricao.push(new ArquivoPrescricaoDTO().jsonToEntity(arquivo));
			});
		}

		if (res.arquivosPrescricaoJustificativa) {
			this.arquivosPrescricaoJustificativa = [] ;
			res.arquivosPrescricaoJustificativa.forEach(arquivo => {
				this.arquivosPrescricaoJustificativa.push(new ArquivoPrescricaoDTO().jsonToEntity(arquivo));
			});
		}

		if (res.medula) {
			this.medula = new PrescricaoMedulaDTO().jsonToEntity(res.medula);
		}

      if (res.cordao) {
			this.cordao = new PrescricaoCordaoDTO().jsonToEntity(res.cordao);
		}

      this.id = ConvertUtil.parseJsonParaAtributos(res.id, new Number());
      this.idTipoPrescricao = ConvertUtil.parseJsonParaAtributos(res.idTipoPrescricao, new Number());
      this.idDoador = ConvertUtil.parseJsonParaAtributos(res.idDoador, new Number());
      this.idTipoDoador = ConvertUtil.parseJsonParaAtributos(res.idTipoDoador, new Number());
      this.rmr = ConvertUtil.parseJsonParaAtributos(res.rmr, new Number());
      this.idCentroTransplante = ConvertUtil.parseJsonParaAtributos(res.idCentroTransplante, new Number());

		return this;
	}

}

File: redome-workup/src/main/java/br/org/cancer/redome/workup/dto/PrescricaoDTO.java
package br.org.cancer.redome.workup.dto;

import java.util.List;
import java.util.stream.Collectors;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

import br.org.cancer.redome.workup.model.Prescricao;
import br.org.cancer.redome.workup.model.PrescricaoCordao;
import br.org.cancer.redome.workup.model.PrescricaoMedula;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * DTO para prescrição.
 * 
 * @author ergomes
 *
 */
@NoArgsConstructor
@AllArgsConstructor
@Data
@Builder
@JsonIgnoreProperties(ignoreUnknown = true)
public class PrescricaoDTO {

	private Long id;

	private Long idTipoPrescricao;
	
	private Long idDoador;

	private Long idTipoDoador;
	
	private Long rmr;

	private Long idCentroTransplante;
	
	private PrescricaoMedulaDTO medula;
	
	private PrescricaoCordaoDTO cordao;
	
	private List<ArquivoPrescricaoDTO> arquivosPrescricao;

	private List<ArquivoPrescricaoDTO> arquivosPrescricaoJustificativa;
	
	
	public static PrescricaoDTO parse(Prescricao prescricao, SolicitacaoDTO solicitacao) {

		PrescricaoDTO prescricaoDTO = null;
		
		List<ArquivoPrescricaoDTO> arquivosPrescricao = prescricao.getArquivosPrescricao().stream()
				.filter(arquivo -> !arquivo.getJustificativa())
				.map(ArquivoPrescricaoDTO::new).collect(Collectors.toList());

		List<ArquivoPrescricaoDTO> arquivosPrescricaoJustificativa = prescricao.getArquivosPrescricao().stream()
				.filter(arquivo -> arquivo.getJustificativa())
				.map(ArquivoPrescricaoDTO::new).collect(Collectors.toList());
		
		if(prescricao.isPrescricaoMedula()) {

			PrescricaoMedula prescricaoMedula = (PrescricaoMedula) prescricao;
			
			PrescricaoMedulaDTO retornoMedula = PrescricaoMedulaDTO.parse(prescricaoMedula);

			prescricaoDTO = PrescricaoDTO.builder()
					.id(prescricao.getId())
					.idTipoPrescricao(prescricao.getTipoPrescricao())
					.idDoador(solicitacao.getMatch().getDoador().getId())
					.idTipoDoador(solicitacao.getMatch().getDoador().getIdTipoDoador())
					.rmr(solicitacao.getMatch().getBusca().getPaciente().getRmr())
					.medula(retornoMedula)
					.arquivosPrescricao(arquivosPrescricao)
					.arquivosPrescricaoJustificativa(arquivosPrescricaoJustificativa)
					.build();
		}else {
		
			PrescricaoCordao prescricaoCordao = (PrescricaoCordao) prescricao;

			PrescricaoCordaoDTO retornoCordao = PrescricaoCordaoDTO.parse(prescricaoCordao);

			prescricaoDTO = PrescricaoDTO.builder()
					.id(prescricao.getId())
					.idTipoPrescricao(prescricao.getTipoPrescricao())
					.idDoador(solicitacao.getMatch().getDoador().getId())
					.idTipoDoador(solicitacao.getMatch().getDoador().getIdTipoDoador())
					.rmr(solicitacao.getMatch().getBusca().getPaciente().getRmr())
					.idCentroTransplante(solicitacao.getMatch().getBusca().getCentroTransplante().getId())
					.cordao(retornoCordao)
					.arquivosPrescricao(arquivosPrescricao)
					.arquivosPrescricaoJustificativa(arquivosPrescricaoJustificativa)
					.build();
		}
		return prescricaoDTO;
	}
}
